src/flatten.rs

```diff
@@ -0,0 +1,561 @@
+        // Performs DFS to collect all dependencies of a target
+        fn collect_deps(
+            path: &PathBuf,
+            paths: &ProjectPathsConfig,
+            graph: &Graph,
+            deps: &mut HashSet<PathBuf>,
+            ordered_deps: &mut Vec<PathBuf>,
+        ) -> Result<()> {
+            if deps.insert(path.clone()) {
+                let target_dir = path.parent().ok_or_else(|| {
+                    SolcError::msg(format!(
+                        "failed to get parent directory for \"{}\"",
+                        path.display()
+                    ))
+                })?;
+
+                let node_id = graph.files().get(path).ok_or_else(|| {
+                    SolcError::msg(format!("cannot resolve file at {}", path.display()))
+                })?;
+
+                let mut imports = graph.node(*node_id).imports().clone();
+                imports.sort_by_key(|i| i.loc().start);
+
+                for import in imports {
+                    let path = paths.resolve_import(target_dir, import.data().path())?;
+                    collect_deps(&path, paths, graph, deps, ordered_deps)?;
+                }
+                ordered_deps.push(path.clone());
+            }
+            Ok(())
+        }
+
+        let graph = Graph::resolve(&project.paths)?;
+
+        let mut ordered_deps = Vec::new();
+        collect_deps(
+            &target.into(),
+            &project.paths,
+            &graph,
+            &mut HashSet::new(),
+            &mut ordered_deps,
+        )?;
+
+        let sources = Source::read_all(&ordered_deps)?;
+
+        // Convert all ASTs from artifacts to strongly typed ASTs
+        let mut asts: Vec<(PathBuf, SourceUnit)> = Vec::new();
+        for (path, ast) in output.artifacts_with_files().filter_map(|(path, _, artifact)| {
+            if let Some(ast) = artifact.ast.as_ref() {
+                if sources.contains_key(&PathBuf::from(path)) {
+                    Some((path, ast))
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        }) {
+            asts.push((PathBuf::from(path), serde_json::from_str(&serde_json::to_string(ast)?)?));
+        }
+
+        Ok(Flattener { target: target.into(), sources, asts, ordered_sources: ordered_deps })
+    }
+
+    /// Flattens target file and returns the result as a string
+    ///
+    /// Flattening process includes following steps:
+    /// 1. Find all file-level definitions and rename references to them via aliased or qualified
+    ///    imports.
+    /// 2. Find all duplicates among file-level definitions and rename them to avoid conflicts.
+    /// 3. Remove all imports.
+    /// 4. Remove all pragmas except for the ones in the target file.
+    /// 5. Remove all license identifiers except for the one in the target file.
+    pub fn flatten(&self) -> String {
```

Comment:
can we split this into a two step process,

one additional (internal) struct that wraps Updates + Flattener, basically `let updated_sources = self.get_updated_sources(updates)`

that then has function to construct the output string. this could perhaps come in handy for some tests